File: KeeperLegacyTests/BreedingTests.swift
import XCTest
@testable import KeeperLegacy

// MARK: - Breeding Tests
// Tests for breeding cost, genetic inheritance, feature gating, and breed outcomes.

@MainActor
final class BreedingTests: XCTestCase {

    // MARK: Helpers

    private func makeProgressVM(coins: Int = 10_000, breedingUnlocked: Bool = true) -> ProgressionViewModel {
        let vm = ProgressionViewModel()
        vm.coins = coins
        if breedingUnlocked {
            vm.unlockedFeatures.insert(GameFeature.breeding.rawValue)
        }
        return vm
    }

    private func makeAdultCreature(dm: DataManager, ofType type: HabitatType, mutation: Int = 0) -> CreatureEntity {
        let entry   = CreatureCatalogEntry.creatures(ofType: type).first!
        let entity  = dm.createCreature(catalogID: entry.id, mutationIndex: mutation)
        entity.lifecycle = LifecycleStage.adult.rawValue
        return entity
    }

    // MARK: Breeding Cost

    func testBreedingCostByRarity() {
        let commonCost   = PricingTable.Breeding.cost(rarity: .common)
        let uncommonCost = PricingTable.Breeding.cost(rarity: .uncommon)
        let rareCost     = PricingTable.Breeding.cost(rarity: .rare)

        XCTAssertGreaterThan(commonCost,   0,            "Breeding cost must be positive")
        XCTAssertGreaterThan(uncommonCost, commonCost,   "Uncommon breed costs more than common")
        XCTAssertGreaterThan(rareCost,     uncommonCost, "Rare breed costs more than uncommon")
    }

    func testBreedingCostIsPositiveForAllRarities() {
        for rarity in Rarity.allCases {
            XCTAssertGreaterThan(PricingTable.Breeding.cost(rarity: rarity), 0,
                                 "\(rarity) breeding cost must be positive")
        }
    }

    // MARK: Feature Gating

    func testBreedingRequiresLevel12() {
        XCTAssertEqual(GameFeature.breeding.requiredLevel, 12,
                       "Breeding should unlock at Level 12 per design spec")
    }

    func testBreedingNotUnlockedAtLevel1() {
        var progression = PlayerProgression()
        progression.currentLevel = 1
        XCTAssertFalse(progression.unlockedFeatures.contains(GameFeature.breeding.rawValue),
                       "Breeding should not be available at level 1")
    }

    func testBreedingUnlocksAtLevel12() {
        var progression          = PlayerProgression()
        progression.currentLevel = 1
        progression.currentXP    = 0
        progression.unlockedFeatures = ["shop", "habitat", "feeding", "playing"]

        // Pump XP until level 12
        while progression.currentLevel < 12 {
            let xpNeeded = XPCurve.xpRequired(forLevel: progression.currentLevel)
            _ = progression.addXP(xpNeeded + 1)
        }

        XCTAssertGreaterThanOrEqual(progression.currentLevel, 12)
        XCTAssertTrue(progression.unlockedFeatures.contains(GameFeature.breeding.rawValue),
                      "Breeding should be in unlockedFeatures after reaching level 12")
    }

    // MARK: Mutation Probability Weights

    func testMutationWeightsSumToOne() {
        let weights: [Double] = [0.40, 0.30, 0.20, 0.10]
        let total = weights.reduce(0.0, +)
        XCTAssertEqual(total, 1.0, accuracy: 0.0001,
                       "Mutation probability weights must sum to exactly 1.0")
    }

    func testMutationWeightsDecreaseByIndex() {
        let weights: [Double] = [0.40, 0.30, 0.20, 0.10]
        for i in 0..<(weights.count - 1) {
            XCTAssertGreaterThan(weights[i], weights[i + 1],
                                 "Higher mutation indices should have lower probability")
        }
    }

    func testAllCreaturesHaveFourMutations() {
        for entry in CreatureCatalogEntry.allCreatures {
            XCTAssertEqual(entry.mutations.count, 4,
                           "\(entry.name) must have exactly 4 mutation variants")
        }
    }

    // MARK: Species Compatibility

    func testSameHabitatTypeCreaturesAreCompatible() {
        let waterCreatures = CreatureCatalogEntry.creatures(ofType: .water)
        XCTAssertGreaterThanOrEqual(waterCreatures.count, 2,
                                    "Need at least 2 water creatures to test breeding")
        XCTAssertEqual(waterCreatures[0].habitatType, waterCreatures[1].habitatType)
    }

    func testDifferentHabitatTypesAreIncompatible() {
        let water = CreatureCatalogEntry.creatures(ofType: .water).first!
        let fire  = CreatureCatalogEntry.creatures(ofType: .fire).first!
        XCTAssertNotEqual(water.habitatType, fire.habitatType)
    }

    // MARK: Breed Outcomes (in-memory Core Data)

    func testBreedProducesOffspringWithValidCatalogID() async {
        let dm      = DataManager.preview
        let parentA = makeAdultCreature(dm: dm, ofType: .water, mutation: 0)
        let parentB = makeAdultCreature(dm: dm, ofType: .water, mutation: 1)

        let entryA = CreatureCatalogEntry.find(byID: parentA.catalogID!)!
        let entryB = CreatureCatalogEntry.find(byID: parentB.catalogID!)!

        let progressVM  = makeProgressVM()
        let creatureVM  = CreatureViewModel()
        creatureVM.breed(parentA: parentA, parentB: parentB, progressVM: progressVM)

        guard case let .success(offspringCatalogID, mutationIndex) = creatureVM.breedingResult else {
            XCTFail("Expected breeding success")
            return
        }

        XCTAssertTrue([entryA.id, entryB.id].contains(offspringCatalogID),
                      "Offspring catalogID must be one of the two parent species")
        XCTAssertGreaterThanOrEqual(mutationIndex, 0)
        XCTAssertLessThanOrEqual(mutationIndex,    3)
    }

    func testBreedDeductsCoins() {
        let dm      = DataManager.preview
        let parentA = makeAdultCreature(dm: dm, ofType: .grass)
        let parentB = makeAdultCreature(dm: dm, ofType: .grass)

        let entryA      = CreatureCatalogEntry.find(byID: parentA.catalogID!)!
        let expectedCost = PricingTable.Breeding.cost(rarity: entryA.rarity)

        let progressVM = makeProgressVM(coins: expectedCost + 500)
        let balanceBefore = progressVM.coins

        let creatureVM = CreatureViewModel()
        creatureVM.breed(parentA: parentA, parentB: parentB, progressVM: progressVM)

        XCTAssertEqual(progressVM.coins, balanceBefore - expectedCost,
                       "Breeding should deduct exactly the rarity-based cost")
    }

    func testBreedFailsWhenInsufficientCoins() {
        let dm      = DataManager.preview
        let parentA = makeAdultCreature(dm: dm, ofType: .fire)
        let parentB = makeAdultCreature(dm: dm, ofType: .fire)

        let progressVM = makeProgressVM(coins: 0)

        let creatureVM = CreatureViewModel()
        creatureVM.breed(parentA: parentA, parentB: parentB, progressVM: progressVM)

        guard case .failure = creatureVM.breedingResult else {
            XCTFail("Expected failure when coins are insufficient")
            return
        }
        XCTAssertEqual(progressVM.coins, 0, "Coins should be unchanged after failed breed")
    }

    func testBreedFailsWhenFeatureNotUnlocked() {
        let dm      = DataManager.preview
        let parentA = makeAdultCreature(dm: dm, ofType: .ice)
        let parentB = makeAdultCreature(dm: dm, ofType: .ice)

        let progressVM = makeProgressVM(breedingUnlocked: false)

        let creatureVM = CreatureViewModel()
        creatureVM.breed(parentA: parentA, parentB: parentB, progressVM: progressVM)

        guard case .failure = creatureVM.breedingResult else {
            XCTFail("Expected failure when breeding is not yet unlocked")
            return
        }
    }

    func testBreedFailsForDifferentHabitatTypes() {
        let dm      = DataManager.preview
        let parentA = makeAdultCreature(dm: dm, ofType: .water)
        let parentB = makeAdultCreature(dm: dm, ofType: .fire)

        let progressVM = makeProgressVM()

        let creatureVM = CreatureViewModel()
        creatureVM.breed(parentA: parentA, parentB: parentB, progressVM: progressVM)

        guard case .failure = creatureVM.breedingResult else {
            XCTFail("Expected failure when parents have different habitat types")
            return
        }
    }

    func testBreedCreatesBreedingRecord() {
        let dm      = DataManager.preview
        let parentA = makeAdultCreature(dm: dm, ofType: .electric)
        let parentB = makeAdultCreature(dm: dm, ofType: .electric)

        let parentAID = parentA.id!
        let parentBID = parentB.id!

        let progressVM = makeProgressVM()
        let creatureVM = CreatureViewModel()
        creatureVM.breed(parentA: parentA, parentB: parentB, progressVM: progressVM)

        guard case .success = creatureVM.breedingResult else {
            XCTFail("Breeding should succeed for lineage test")
            return
        }

        let record = dm.allBreedingRecords().first
        XCTAssertNotNil(record,            "A breeding record should be saved after successful breed")
        XCTAssertEqual(record?.parentAID,  parentAID)
        XCTAssertEqual(record?.parentBID,  parentBID)
        XCTAssertNotNil(record?.offspringID)
    }

    func testBreedAddsNewCreatureToOwnedCount() {
        let dm           = DataManager.preview
        let countBefore  = dm.allOwnedCreatures().count

        let parentA = makeAdultCreature(dm: dm, ofType: .dirt)
        let parentB = makeAdultCreature(dm: dm, ofType: .dirt)

        let progressVM = makeProgressVM()
        let creatureVM = CreatureViewModel()
        creatureVM.breed(parentA: parentA, parentB: parentB, progressVM: progressVM)

        guard case .success = creatureVM.breedingResult else {
            XCTFail("Breeding should succeed")
            return
        }

        // 2 parents were added + 1 offspring = countBefore + 3
        XCTAssertEqual(dm.allOwnedCreatures().count, countBefore + 3,
                       "Two parents and one offspring should be in inventory after breeding")
    }
}

File: KeeperLegacyGodot/Tests/BreedingTests.cs
// Tests/BreedingTests.cs
// Port of BreedingTests.swift
// Tests breeding costs, mutation weights, compatibility rules, and BreedingEngine outcomes.

using System;
using System.Linq;
using NUnit.Framework;
using KeeperLegacy.Data;
using KeeperLegacy.Models;

namespace KeeperLegacy.Tests
{
    [TestFixture]
    public class BreedingTests
    {
        // ── Breeding Costs ────────────────────────────────────────────────────

        [Test]
        public void BreedingCostIsPositiveForAllRarities()
        {
            Assert.That(PricingTable.Breeding.Cost(Rarity.Common),   Is.GreaterThan(0));
            Assert.That(PricingTable.Breeding.Cost(Rarity.Uncommon), Is.GreaterThan(0));
            Assert.That(PricingTable.Breeding.Cost(Rarity.Rare),     Is.GreaterThan(0));
        }

        [Test]
        public void BreedingCostOrderingByRarity()
        {
            int common   = PricingTable.Breeding.Cost(Rarity.Common);
            int uncommon = PricingTable.Breeding.Cost(Rarity.Uncommon);
            int rare     = PricingTable.Breeding.Cost(Rarity.Rare);
            Assert.That(common,   Is.LessThan(uncommon));
            Assert.That(uncommon, Is.LessThan(rare));
        }

        [Test]
        public void BreedingCostUsesLowerRarityParent()
        {
            // Cost is determined by the lower-rarity parent
            int commonCost = PricingTable.Breeding.Cost(Rarity.Common);
            var entryA = CreatureRosterData.Find("aquaburst")!;   // Common
            var entryB = CreatureRosterData.Find("seraphine")!;   // Uncommon water — Common cost applies
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("seraphine");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 500);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            engine.Breed(parentA, parentB, entryA, entryB, economy, progression, new Random(42));

            // 500 − commonCost coins should remain (lower rarity cost applied)
            Assert.That(economy.Coins, Is.EqualTo(500 - commonCost));
        }

        // ── Feature Gating ────────────────────────────────────────────────────

        [Test]
        public void BreedingFailsWhenFeatureNotUnlocked()
        {
            var entry   = CreatureRosterData.Find("aquaburst")!;
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("aquaburst");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1000);
            var progression = new PlayerProgression();  // Level 1 — breeding locked

            var engine = new BreedingEngine();
            var result = engine.Breed(parentA, parentB, entry, entry, economy, progression);

            Assert.That(result, Is.InstanceOf<BreedingEngine.Failure>());
        }

        [Test]
        public void BreedingSucceedsWhenFeatureUnlocked()
        {
            var entry   = CreatureRosterData.Find("aquaburst")!;
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("aquaburst");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1000);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            var result = engine.Breed(parentA, parentB, entry, entry, economy, progression, new Random(1));

            Assert.That(result, Is.InstanceOf<BreedingEngine.Success>());
        }

        // ── Mutation Weights ─────────────────────────────────────────────────

        [Test]
        public void MutationWeightsSumToOne()
        {
            double sum = BreedingEngine.MutationWeights.Sum();
            Assert.That(sum, Is.EqualTo(1.0).Within(0.001));
        }

        [Test]
        public void MutationWeightsDecreaseByIndex()
        {
            var w = BreedingEngine.MutationWeights;
            Assert.That(w[0], Is.GreaterThan(w[1]));
            Assert.That(w[1], Is.GreaterThan(w[2]));
            Assert.That(w[2], Is.GreaterThan(w[3]));
        }

        [Test]
        public void MutationWeightsHaveFourEntries()
            => Assert.That(BreedingEngine.MutationWeights.Length, Is.EqualTo(4));

        [Test]
        public void SelectMutationReturnsValidIndex()
        {
            var engine = new BreedingEngine();
            for (int i = 0; i < 50; i++)
            {
                int idx = engine.SelectMutation(new Random(i));
                Assert.That(idx, Is.InRange(0, 3));
            }
        }

        // ── Habitat Compatibility ─────────────────────────────────────────────

        [Test]
        public void SameHabitatTypeIsCompatible()
        {
            var entryA  = CreatureRosterData.Find("aquaburst")!;   // Water
            var entryB  = CreatureRosterData.Find("tidecaller")!;  // Water
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("tidecaller");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1000);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            var result = engine.Breed(parentA, parentB, entryA, entryB, economy, progression, new Random(0));

            Assert.That(result, Is.InstanceOf<BreedingEngine.Success>());
        }

        [Test]
        public void DifferentHabitatTypesFail()
        {
            var entryA  = CreatureRosterData.Find("aquaburst")!;   // Water
            var entryB  = CreatureRosterData.Find("emberpaw")!;    // Fire
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("emberpaw");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1000);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            var result = engine.Breed(parentA, parentB, entryA, entryB, economy, progression);

            Assert.That(result, Is.InstanceOf<BreedingEngine.Failure>());
        }

        // ── Lifecycle Requirement ─────────────────────────────────────────────

        [Test]
        public void NonAdultParentFails()
        {
            var entry   = CreatureRosterData.Find("aquaburst")!;
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("aquaburst");
            parentA.Lifecycle = LifecycleStage.Baby;    // Explicitly not adult
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1000);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            var result = engine.Breed(parentA, parentB, entry, entry, economy, progression);

            Assert.That(result, Is.InstanceOf<BreedingEngine.Failure>());
        }

        // ── Insufficient Coins ────────────────────────────────────────────────

        [Test]
        public void InsufficientCoinsFails()
        {
            var entry   = CreatureRosterData.Find("aquaburst")!;
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("aquaburst");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1);   // Way too few
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            var result = engine.Breed(parentA, parentB, entry, entry, economy, progression);

            Assert.That(result, Is.InstanceOf<BreedingEngine.Failure>());
            Assert.That(economy.Coins, Is.EqualTo(1));   // Coins not deducted on failure
        }

        // ── Breed Success Outcome ─────────────────────────────────────────────

        [Test]
        public void SuccessOffspringCatalogIdIsParent()
        {
            var entry   = CreatureRosterData.Find("aquaburst")!;
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("aquaburst");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1000);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine  = new BreedingEngine();
            var result  = engine.Breed(parentA, parentB, entry, entry, economy, progression, new Random(5)) as BreedingEngine.Success;

            Assert.That(result, Is.Not.Null);
            Assert.That(result!.OffspringCatalogId, Is.EqualTo("aquaburst"));
        }

        [Test]
        public void SuccessDeductsCoins()
        {
            var entry   = CreatureRosterData.Find("aquaburst")!;
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("aquaburst");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            int startCoins  = 1000;
            var economy     = new PlayerEconomy(startingCoins: startCoins);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            engine.Breed(parentA, parentB, entry, entry, economy, progression, new Random(5));

            Assert.That(economy.Coins, Is.LessThan(startCoins));
        }

        // ── Breeding Record ───────────────────────────────────────────────────

        [Test]
        public void SuccessResultContainsParentIds()
        {
            var entry   = CreatureRosterData.Find("aquaburst")!;
            var parentA = new CreatureInstance("aquaburst");
            var parentB = new CreatureInstance("aquaburst");
            parentA.Lifecycle = LifecycleStage.Adult;
            parentB.Lifecycle = LifecycleStage.Adult;

            var economy     = new PlayerEconomy(startingCoins: 1000);
            var progression = new PlayerProgression();
            progression.CurrentLevel = 12;
            progression.StoryAct     = 1;
            progression.CheckAndUnlockFeatures();

            var engine = new BreedingEngine();
            var result = engine.Breed(parentA, parentB, entry, entry, economy, progression, new Random(5)) as BreedingEngine.Success;

            Assert.That(result, Is.Not.Null);
            Assert.That(result!.Record.ParentAId, Is.EqualTo(parentA.Id));
            Assert.That(result.Record.ParentBId,  Is.EqualTo(parentB.Id));
        }
    }
}
